File: kafka-platform/clients/java/producer-example/src/main/java/com/example/kafka/ProducerExample.java
package com.example.kafka;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.StringSerializer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Instant;
import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

/**
 * Production-ready Kafka producer example with:
 * - Idempotent producer
 * - Exactly-once semantics
 * - Error handling
 * - Metrics tracking
 * - Graceful shutdown
 */
public class ProducerExample {
    private static final Logger logger = LoggerFactory.getLogger(ProducerExample.class);
    private static final ObjectMapper objectMapper = new ObjectMapper();

    private final KafkaProducer<String, String> producer;
    private final String topic;
    private long messagesSent = 0;
    private long errorsCount = 0;

    public ProducerExample(String bootstrapServers, String topic) {
        this.topic = topic;
        this.producer = createProducer(bootstrapServers);
        
        // Graceful shutdown hook
        Runtime.getRuntime().addShutdownHook(new Thread(this::close));
    }

    private KafkaProducer<String, String> createProducer(String bootstrapServers) {
        Properties props = new Properties();
        
        // Connection
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
        
        // Serialization
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        
        // Idempotence & Exactly-once semantics
        props.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, true);
        props.put(ProducerConfig.ACKS_CONFIG, "all"); // Wait for all in-sync replicas
        props.put(ProducerConfig.RETRIES_CONFIG, Integer.MAX_VALUE);
        props.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, 5);
        
        // Compression for network efficiency
        props.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, "snappy");
        
        // Batching for throughput
        props.put(ProducerConfig.BATCH_SIZE_CONFIG, 32768); // 32 KB
        props.put(ProducerConfig.LINGER_MS_CONFIG, 10);
        
        // Buffer settings
        props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, 33554432); // 32 MB
        
        // Request timeout
        props.put(ProducerConfig.REQUEST_TIMEOUT_MS_CONFIG, 30000);
        props.put(ProducerConfig.DELIVERY_TIMEOUT_MS_CONFIG, 120000);
        
        // Client ID for tracking
        props.put(ProducerConfig.CLIENT_ID_CONFIG, "producer-example-" + UUID.randomUUID());
        
        // Security (uncomment and configure for production)
        // props.put("security.protocol", "SASL_SSL");
        // props.put("sasl.mechanism", "SCRAM-SHA-512");
        // props.put("sasl.jaas.config", 
        //     "org.apache.kafka.common.security.scram.ScramLoginModule required " +
        //     "username=\"producer-user\" password=\"secure-password\";");
        // props.put("ssl.truststore.location", "/path/to/truststore.jks");
        // props.put("ssl.truststore.password", "truststore-password");
        
        logger.info("Creating Kafka producer with idempotence enabled");
        return new KafkaProducer<>(props);
    }

    /**
     * Send a message synchronously (blocking).
     * Use for scenarios where you need immediate confirmation.
     */
    public void sendSync(String key, Map<String, Object> value) throws Exception {
        String jsonValue = objectMapper.writeValueAsString(value);
        ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, jsonValue);
        
        try {
            RecordMetadata metadata = producer.send(record).get();
            messagesSent++;
            logger.info("Message sent successfully - Topic: {}, Partition: {}, Offset: {}, Key: {}",
                    metadata.topic(), metadata.partition(), metadata.offset(), key);
        } catch (InterruptedException | ExecutionException e) {
            errorsCount++;
            logger.error("Failed to send message with key: {}", key, e);
            throw e;
        }
    }

    /**
     * Send a message asynchronously (non-blocking).
     * Use for high-throughput scenarios.
     */
    public void sendAsync(String key, Map<String, Object> value) throws Exception {
        String jsonValue = objectMapper.writeValueAsString(value);
        ProducerRecord<String, String> record = new ProducerRecord<>(topic, key, jsonValue);
        
        producer.send(record, new Callback() {
            @Override
            public void onCompletion(RecordMetadata metadata, Exception exception) {
                if (exception == null) {
                    messagesSent++;
                    logger.debug("Message sent - Partition: {}, Offset: {}", 
                            metadata.partition(), metadata.offset());
                } else {
                    errorsCount++;
                    logger.error("Failed to send message with key: {}", key, exception);
                }
            }
        });
    }

    /**
     * Send a batch of messages efficiently.
     */
    public void sendBatch(List<Map<String, Object>> messages) throws Exception {
        for (int i = 0; i < messages.size(); i++) {
            String key = "batch-" + i;
            sendAsync(key, messages.get(i));
        }
        
        // Flush to ensure all messages are sent
        producer.flush();
        logger.info("Batch of {} messages sent", messages.size());
    }

    public void close() {
        logger.info("Closing producer. Stats - Sent: {}, Errors: {}", messagesSent, errorsCount);
        producer.close();
    }

    public static void main(String[] args) throws Exception {
        // Configuration
        String bootstrapServers = System.getenv().getOrDefault("KAFKA_BOOTSTRAP", "localhost:9092");
        String topic = System.getenv().getOrDefault("KAFKA_TOPIC", "orders");
        
        ProducerExample producer = new ProducerExample(bootstrapServers, topic);
        
        try {
            // Example 1: Send individual messages synchronously
            logger.info("Sending individual messages synchronously...");
            for (int i = 0; i < 10; i++) {
                Map<String, Object> order = new HashMap<>();
                order.put("order_id", "ORD-" + String.format("%04d", i));
                order.put("customer_id", "CUST-" + (i % 5));
                order.put("amount", 99.99 + i);
                order.put("timestamp", Instant.now().toEpochMilli());
                
                producer.sendSync("order-" + i, order);
                Thread.sleep(100); // Simulate processing delay
            }
            
            // Example 2: Send batch of messages asynchronously
            logger.info("Sending batch of messages asynchronously...");
            List<Map<String, Object>> batch = new ArrayList<>();
            for (int i = 0; i < 100; i++) {
                Map<String, Object> event = new HashMap<>();
                event.put("event_id", "EVT-" + i);
                event.put("type", "page_view");
                event.put("user_id", "USER-" + (i % 10));
                event.put("timestamp", Instant.now().toEpochMilli());
                batch.add(event);
            }
            producer.sendBatch(batch);
            
            logger.info("All messages sent successfully!");
            
        } catch (Exception e) {
            logger.error("Error in main execution", e);
            throw e;
        } finally {
            producer.close();
        }
    }
}

File: kafka-platform/clients/python/producer_example.py
"""
Kafka Producer Example with Best Practices
- Idempotent producer
- Compression
- Error handling
- Metrics
"""

import json
import time
from typing import Dict, Any
from kafka import KafkaProducer
from kafka.errors import KafkaError
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class KafkaProducerExample:
    def __init__(self, bootstrap_servers: str, topic: str):
        """
        Initialize Kafka producer with production-ready settings.
        
        Args:
            bootstrap_servers: Comma-separated list of broker addresses
            topic: Topic to produce to
        """
        self.topic = topic
        self.producer = KafkaProducer(
            bootstrap_servers=bootstrap_servers.split(','),
            
            # Serialization
            value_serializer=lambda v: json.dumps(v).encode('utf-8'),
            key_serializer=lambda k: k.encode('utf-8') if k else None,
            
            # Idempotence & Exactly-once semantics
            enable_idempotence=True,
            acks='all',  # Wait for all in-sync replicas
            retries=3,
            max_in_flight_requests_per_connection=5,
            
            # Compression for network efficiency
            compression_type='snappy',
            
            # Batching for throughput
            batch_size=16384,
            linger_ms=10,
            
            # Buffer settings
            buffer_memory=33554432,  # 32MB
            
            # Request timeout
            request_timeout_ms=30000,
            
            # Security (uncomment and configure for production)
            # security_protocol='SASL_SSL',
            # sasl_mechanism='SCRAM-SHA-512',
            # sasl_plain_username='producer-user',
            # sasl_plain_password='secure-password',
            # ssl_cafile='/path/to/ca-cert',
        )
        
        self.messages_sent = 0
        self.errors = 0
        
    def send_message(self, key: str, value: Dict[str, Any]) -> None:
        """
        Send a message to Kafka with error handling.
        
        Args:
            key: Message key for partitioning
            value: Message value (will be JSON serialized)
        """
        try:
            future = self.producer.send(
                self.topic,
                key=key,
                value=value
            )
            
            # Wait for send to complete (synchronous for example)
            # In production, use async callbacks for better throughput
            record_metadata = future.get(timeout=10)
            
            self.messages_sent += 1
            logger.info(
                f"Message sent successfully: "
                f"topic={record_metadata.topic}, "
                f"partition={record_metadata.partition}, "
                f"offset={record_metadata.offset}"
            )
            
        except KafkaError as e:
            self.errors += 1
            logger.error(f"Failed to send message: {e}")
            raise
    
    def send_batch(self, messages: list) -> None:
        """
        Send multiple messages efficiently.
        
        Args:
            messages: List of (key, value) tuples
        """
        for key, value in messages:
            try:
                # Async send for better throughput
                self.producer.send(self.topic, key=key, value=value)
            except Exception as e:
                logger.error(f"Error queuing message: {e}")
        
        # Flush all pending messages
        self.producer.flush()
        logger.info(f"Batch of {len(messages)} messages sent")
    
    def close(self):
        """Gracefully close the producer."""
        logger.info(f"Closing producer. Stats: sent={self.messages_sent}, errors={self.errors}")
        self.producer.close()


def main():
    """Example usage"""
    bootstrap_servers = "localhost:9092"
    topic = "orders"
    
    producer = KafkaProducerExample(bootstrap_servers, topic)
    
    try:
        # Send individual messages
        for i in range(10):
            order = {
                "order_id": f"ORD-{i:04d}",
                "customer_id": f"CUST-{i % 5}",
                "amount": 99.99 + i,
                "timestamp": int(time.time() * 1000)
            }
            producer.send_message(key=f"order-{i}", value=order)
            time.sleep(0.1)
        
        # Send batch
        batch_messages = [
            (f"batch-{i}", {"batch_id": i, "data": f"value-{i}"})
            for i in range(100)
        ]
        producer.send_batch(batch_messages)
        
    finally:
        producer.close()


if __name__ == "__main__":
    main()
